python/postprocessing/modules/jme/jetSmearer.py

```diff
@@ -25,52 +25,63 @@ def __init__(self, globalTag, jetType = "AK4PFchs", jerInputFileName = "Spring16
         self.jerArchive.extractall(self.jerInputFilePath)
         self.jerInputFileName = jerInputFileName
         self.jerUncertaintyInputFileName = jerUncertaintyInputFileName
-
+        
         self.jmr_vals = jmr_vals
-
+        
         self.params_sf_and_uncertainty = ROOT.PyJetParametersWrapper()
         self.params_resolution = ROOT.PyJetParametersWrapper()
-
+        
         # initialize random number generator
         # (needed for jet pT smearing)
         self.rnd = ROOT.TRandom3(12345)
-
+        
         # load libraries for accessing JER scale factors and uncertainties from txt files
         for library in [ "libCondFormatsJetMETObjects", "libPhysicsToolsNanoAODTools" ]:
             if library not in ROOT.gSystem.GetLibraries():
                 print("Load Library '%s'" % library.replace("lib", ""))
                 ROOT.gSystem.Load(library)
-
+        
         self.puppiJMRFile = ROOT.TFile.Open(os.environ['CMSSW_BASE'] + "/src/PhysicsTools/NanoAODTools/data/jme/puppiSoftdropResol.root")
         self.puppisd_resolution_cen = self.puppiJMRFile.Get("massResolution_0eta1v3")
         self.puppisd_resolution_for = self.puppiJMRFile.Get("massResolution_1v3eta2v5")
-
+        
     def beginJob(self):
-
         # initialize JER scale factors and uncertainties
         # (cf. PhysicsTools/PatUtils/interface/SmearedJetProducerT.h )
         print("Loading jet energy resolutions (JER) from file '%s'" % os.path.join(self.jerInputFilePath, self.jerInputFileName))
         self.jer = ROOT.PyJetResolutionWrapper(os.path.join(self.jerInputFilePath, self.jerInputFileName))
         print("Loading JER scale factors and uncertainties from file '%s'" % os.path.join(self.jerInputFilePath, self.jerUncertaintyInputFileName))
         self.jerSF_and_Uncertainty = ROOT.PyJetResolutionScaleFactorWrapper(os.path.join(self.jerInputFilePath, self.jerUncertaintyInputFileName))
-
+        
     def endJob(self):
         pass
-
         
+    
+    def setSeed(self,event):
```

Comment:
In [`PhysicsTools/PatUtils/interface/SmearedJetProducerT.h`](https://github.com/cms-sw/cmssw/blob/1fef057ea0118be780f3ccd15eb7b1ce7c9a9dab/PhysicsTools/PatUtils/interface/SmearedJetProducerT.h#L208-L215) that you referred to in the comments, the run number is shifted by 20 bits, lumi number by 10 bits and event number is not shifted at all. I think this is because typically the run number is the smallest, followed by lumi and event number. Shifting them as it's done in C++ code protects the resulting seed from overflowing. Even though the limit on the integer type in Python is ridiculously high, I would still prefer that you implement the same logic of shifting to make the C++ and the Python code equivalent.

One thing I implemented differently wrt the C++ version of the code is that I obtained a different seed not only for every event, but also for every single jet. This is because picking the first jet (by corrected pT) in the event is subject to the effects of JEC, as the order of jets depends on the choice of GT in NanoAOD production.

Reply:
Thank, was not paying attention; I commited a change, such that only the run number and lumi block are shifted, not the event number.

I was also think that the order of jets might be an issue. Now two independent users should be able to compare there jets one-to-one, as long as they start from the same JECs (and therefore the same jet order). Do you think it's still better to do it differently than it is now? Where can I see your implementation, is it possible it's private on GitHub?

Reply:
Thanks for prompt reply and quick fix in the code.

I have no strong opinion about the seed computation for every single jet, as the effect is negligible. On the other hand, it's not incorrect, computationally expensive or difficult to implement either, so it's up to you if you want to keep the code as it is now, or if you follow my proposal. Here is my implementation for reference:

- run-lumi-event numbers as passed to `getSmearValsPt` and to `getSmearValsPt` functions ([here](https://github.com/HEP-KBFI/nanoAOD-tools/blob/185d56ed23b99a4d86122ecf6f92038d40afa95e/python/postprocessing/modules/jme/jetSmearer.py#L74) and [here](https://github.com/HEP-KBFI/nanoAOD-tools/blob/185d56ed23b99a4d86122ecf6f92038d40afa95e/python/postprocessing/modules/jme/jetSmearer.py#L153)) where the seeds are computed
- the function that computes the seed is [here](https://github.com/HEP-KBFI/nanoAOD-tools/blob/185d56ed23b99a4d86122ecf6f92038d40afa95e/python/postprocessing/modules/jme/jetSmearer.py#L10-L20)

I guess it can be optimized by passing the run-lumi-event numbers first and the call `getSmearVals*` without the run-lumi-event argument. (Once this PR gets merged I'll drop whatever I have implemented in my fork in favor of the implementation in upstream.)

Reply:
Using each jet's eta, like in your implementation makes sense, but if you have no strong opinion to include this, I propose to just leave it as it is now. :-p